File: backend/app/schemas/investigation.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field

# Base Enums/Literals
InvestigationStatusEnum = str # CREATED, ANALYZING, COMPLETED, REVIEW_REQUIRED, INSUFFICIENT_EVIDENCE, FAILED
ConfidenceStatusEnum = str # HIGH CONFIDENCE, SUPPORTED, AMBIGUOUS, CONFLICTING, INSUFFICIENT EVIDENCE, UNDER REVIEW

class InvestigationCreate(BaseModel):
    subject_name: str
    alias: Optional[str] = ""
    organization: Optional[str] = ""
    known_platform: Optional[str] = ""
    additional_context: Optional[str] = ""
    image_reference: Optional[str] = None
    consent_confirmed: bool = True

class SignalCheckSchema(BaseModel):
    name: str
    matched: bool
    status: str # match, conflict, unverified, ambiguous
    detail: str

class PlatformSchema(BaseModel):
    platform: str
    handle: str
    verified: bool = False
    url: Optional[str] = None

class CandidateSchema(BaseModel):
    id: str
    candidate_code: str
    name: str
    username: str
    avatar_url: Optional[str] = None
    primary_role: Optional[str] = None
    organizations: List[str] = []
    location: Optional[str] = None
    platforms: List[PlatformSchema] = []
    status: str # SUPPORTED, AMBIGUOUS, CONFLICTING, INSUFFICIENT EVIDENCE
    status_note: str
    evidence_count: int = 0
    supporting_signals: List[str] = []
    conflicting_signals: List[str] = []
    signal_breakdown: List[SignalCheckSchema] = []
    match_score_explanation: str
    # Evidence Adversarial Fields
    supporting_evidence: List[Dict[str, Any]] = []
    contradicting_evidence: List[Dict[str, Any]] = []
    evidence_strength: float = 0.0
    major_concerns: List[str] = []
    final_evidence_status: Optional[str] = None

class SignalMetricSchema(BaseModel):
    name: str
    value: int # 0-100
    label: str # Supported, Strong, Consistent, Partial, Weak, Conflicting
    description: str
    supporting_evidence_count: Optional[int] = 1
    contradicting_evidence_count: Optional[int] = 0
    details: Optional[str] = None

class PipelineStepSchema(BaseModel):
    id: str
    name: str
    short_name: str
    status: str # completed, current, warning, conflict, pending
    description: str
    timestamp: Optional[str] = None

class CounterfactualItemSchema(BaseModel):
    id: str
    title: str
    category: str
    detail: str
    source: str
    confidence_weight: int

class CounterfactualSchema(BaseModel):
    overall_assessment: str
    assessment_rationale: str
    supporting: List[CounterfactualItemSchema] = []
    contradicting: List[CounterfactualItemSchema] = []

class TwinGuardCandidate(BaseModel):
    name: str
    handle: str
    org: str
    location: str
    avatar: str
    key_distinctive_factor: str

class TwinGuardSchema(BaseModel):
    detected: bool
    severity: str # low, medium, high
    title: str
    subtitle: str
    reasons: List[str] = []
    candidate_a: TwinGuardCandidate
    candidate_b: TwinGuardCandidate
    recommendation: str
    required_action: str

class TimelineItemSchema(BaseModel):
    id: str
    year: str
    date_range: Optional[str] = None
    role: str
    organization: str
    category: str # Career, Education, Projects, Events, Publications, Organizations
    evidence_text: str
    source: str
    source_reliability: str = "HIGH"
    is_conflict: bool = False
    conflict_details: Optional[str] = None
    verified: bool = True

class EvidenceSchema(BaseModel):
    id: str
    claim: str
    source: str
    source_type: str
    evidence_snippet: str
    reliability: str # HIGH, MEDIUM, LOW
    timestamp: str
    confidence_level: str
    provenance: str
    supporting_signals: List[str] = []
    conflicting_evidence: List[str] = []
    url: Optional[str] = None

class SourceSchema(BaseModel):
    id: str
    source: str
    platform: str
    type: str # Professional, Technical, Company, Event, Publication, Media
    reliability: str # HIGH, MEDIUM, LOW
    last_checked: str
    evidence_count: int
    status: str # VERIFIED, CONFLICTING, UNRESOLVED
    endpoint_or_domain: str
    verified_signatures: int = 0

class GraphNodeSchema(BaseModel):
    id: str
    type: str # person, organization, project, profile, publication, event, source
    label: str
    subtitle: Optional[str] = None
    avatar: Optional[str] = None
    icon: Optional[str] = None
    status: str = "verified"
    reliability: str = "HIGH"
    claim_count: Optional[int] = None
    metadata: Optional[Dict[str, Any]] = None

class GraphEdgeSchema(BaseModel):
    id: str
    source: str
    target: str
    label: str # works_at, created, attended, contributed_to, published, profile_of, associated_with
    status: str = "verified"
    confidence: Optional[str] = "95%"
    source_ref: Optional[str] = None

class GraphSchema(BaseModel):
    nodes: List[GraphNodeSchema] = []
    edges: List[GraphEdgeSchema] = []

class ConflictSchema(BaseModel):
    id: str
    type: str
    severity: str
    description: str
    sources: List[str] = []

class AIAnalystSchema(BaseModel):
    identity_assessment: str
    key_evidence: List[str] = []
    conflicting_evidence: List[str] = []
    unknowns: List[str] = []
    human_review_recommendation: str
    generated_at: str
    model_provenance: str

class CoverageItemSchema(BaseModel):
    category: str
    coverage_percent: int
    evidence_count: int
    note: str

# Audit Trail Replay Schema
class AuditTrailStepSchema(BaseModel):
    id: str
    step_index: int
    stage: str
    agent: str
    agent_role: str
    action: str
    source: str
    claim: str
    result: str
    timestamp: str
    status: str
    provenance: Optional[str] = None

# Multi-Agent Schema
class AgentRecordSchema(BaseModel):
    id: str
    name: str
    role: str
    permissions: List[str] = []
    constraints: str
    avatar_icon: Optional[str] = None
    status: str = "ACTIVE"
    actions_performed: List[str] = []
    findings_count: int = 0

# Risk & Uncertainty Panel Schema
class RiskUncertaintySchema(BaseModel):
    evidence_status: str # SUPPORTED, AMBIGUOUS, CONFLICTING, INSUFFICIENT EVIDENCE
    supporting_evidence_count: int = 0
    contradicting_evidence_count: int = 0
    unresolved_signals_count: int = 0
    major_concerns: List[str] = []
    human_review_required: bool = True
    risk_level: str = "LOW" # LOW, MEDIUM, HIGH, CRITICAL

# Human-in-the-Loop Review Submission
class HumanReviewSubmission(BaseModel):
    investigation_id: str
    analyst_name: str
    analyst_decision: str # SUPPORTED, AMBIGUOUS, CONFLICTING, INSUFFICIENT EVIDENCE
    accepted_evidence_ids: List[str] = []
    rejected_evidence_ids: List[str] = []
    analyst_notes: str = ""
    sign_off_timestamp: Optional[str] = None

class InvestigationResponse(BaseModel):
    id: str
    subject_name: str
    alias: str
    organization: str
    known_platform: str
    avatar_url: Optional[str] = None
    created_at: str
    status: str
    sources_count: int = 0
    candidates_count: int = 0
    evidence_links_count: int = 0
    ambiguities_count: int = 0
    coverage: List[CoverageItemSchema] = []
    pipeline: List[PipelineStepSchema] = []
    dna_signals: List[SignalMetricSchema] = []
    candidates: List[CandidateSchema] = []
    twin_guard: Optional[TwinGuardSchema] = None
    counterfactual: Optional[CounterfactualSchema] = None
    timeline: List[TimelineItemSchema] = []
    ai_analyst: Optional[AIAnalystSchema] = None
    nodes: List[GraphNodeSchema] = []
    edges: List[GraphEdgeSchema] = []
    # Advanced Investigation Additions
    audit_trail: List[AuditTrailStepSchema] = []
    agents: List[AgentRecordSchema] = []
    risk_uncertainty: Optional[RiskUncertaintySchema] = None
    human_review: Optional[Dict[str, Any]] = None

class PublicProfileSchema(BaseModel):
    id: str
    platform: str
    profile_url: Optional[str] = None
    display_name: Optional[str] = None
    username: Optional[str] = None
    description: Optional[str] = None
    organization: Optional[str] = None
    source_id: Optional[str] = None
    verified: bool = False
    public_metadata: Dict[str, Any] = {}

class AnalysisProgressSchema(BaseModel):
    investigation_id: str
    status: str # IN_PROGRESS, COMPLETED, FAILED, NOT_FOUND
    current_stage: str
    progress_percent: int
    completed_stages: List[str] = []
    sources_found: int = 0
    candidates_found: int = 0
    evidence_found: int = 0
    conflicts_found: int = 0
    result: Optional[InvestigationResponse] = None

File: src/types/investigation.ts
export type InvestigationStatus =
  | "ACTIVE"
  | "COMPLETED"
  | "REVIEW REQUIRED"
  | "INSUFFICIENT EVIDENCE"
  | "ANALYSIS COMPLETE"
  | "HUMAN SIGNED OFF"
  | string;

export type ConfidenceStatus =
  | "HIGH CONFIDENCE"
  | "SUPPORTED"
  | "AMBIGUOUS"
  | "CONFLICTING"
  | "INSUFFICIENT EVIDENCE"
  | "UNDER REVIEW";

export type PipelineStepStatus = "completed" | "current" | "warning" | "conflict" | "pending";

export interface PipelineStep {
  id: string;
  name: string;
  shortName: string;
  status: PipelineStepStatus;
  description: string;
  details?: string;
  timestamp?: string;
}

export interface SignalMetric {
  name: string;
  value: number; // 0 - 100
  label: "Supported" | "Strong" | "Consistent" | "Partial" | "Weak" | "Conflicting" | string;
  description: string;
  supporting_evidence_count?: number;
  contradicting_evidence_count?: number;
  details?: string;
}

export interface SignalCheck {
  name: string;
  matched: boolean;
  status: "match" | "conflict" | "unverified" | "ambiguous";
  detail: string;
}

export interface AdversarialEvidenceItem {
  id: string;
  claim: string;
  category: string;
  source: string;
  url?: string;
  date?: string;
  provenance_hash?: string;
  evidence_type: "SUPPORTING" | "CONTRADICTING" | string;
  weight: number;
  rationale: string;
}

export interface CandidateIdentity {
  id: string;
  candidateCode: "Candidate A" | "Candidate B" | "Candidate C" | "Candidate D" | string;
  name: string;
  username: string;
  avatarUrl: string;
  primaryRole: string;
  organizations: string[];
  platforms: {
    platform: string;
    handle: string;
    verified: boolean;
    url?: string;
  }[];
  status: ConfidenceStatus;
  statusNote: string;
  evidenceCount: number;
  supportingSignals: string[];
  conflictingSignals: string[];
  signalBreakdown: SignalCheck[];
  location?: string;
  matchScoreExplanation: string;
  // Advanced Adversarial Engine
  supporting_evidence?: AdversarialEvidenceItem[];
  contradicting_evidence?: AdversarialEvidenceItem[];
  evidence_strength?: number;
  major_concerns?: string[];
  final_evidence_status?: string;
}

export interface TwinGuardComparison {
  detected: boolean;
  severity: "low" | "medium" | "high";
  title: string;
  subtitle: string;
  reasons: string[];
  candidateA: {
    name: string;
    handle: string;
    org: string;
    location: string;
    avatar: string;
    keyDistinctiveFactor: string;
  };
  candidateB: {
    name: string;
    handle: string;
    org: string;
    location: string;
    avatar: string;
    keyDistinctiveFactor: string;
  };
  recommendation: string;
  requiredAction: string;
}

export interface CounterfactualItem {
  id: string;
  title: string;
  category: "organization" | "username" | "project" | "timeline" | "location" | "role" | "publication" | string;
  detail: string;
  source: string;
  confidenceWeight: number; // 1-5
}

export interface CounterfactualAnalysis {
  supporting: CounterfactualItem[];
  contradicting: CounterfactualItem[];
  overallAssessment: ConfidenceStatus;
  assessmentRationale: string;
}

export interface TimelineEventItem {
  id: string;
  year: string;
  dateRange?: string;
  role: string;
  organization: string;
  category: "Career" | "Education" | "Projects" | "Events" | "Publications" | "Organizations" | string;
  evidenceText: string;
  source: string;
  sourceReliability: "HIGH" | "MEDIUM" | "LOW" | string;
  isConflict?: boolean;
  conflictDetails?: string;
  verified: boolean;
}

export interface AIAnalystReport {
  identityAssessment: string;
  keyEvidence: string[];
  conflictingEvidence: string[];
  unknowns: string[];
  humanReviewRecommendation: string;
  generatedAt: string;
  modelProvenance: string;
}

export type NodeType =
  | "person"
  | "profile"
  | "organization"
  | "project"
  | "event"
  | "publication"
  | "source";

export interface EvidenceNodeData {
  id: string;
  type: NodeType;
  label: string;
  subtitle?: string;
  avatar?: string;
  icon?: string;
  status?: "verified" | "ambiguous" | "conflicting" | "normal";
  reliability?: "HIGH" | "MEDIUM" | "LOW" | string;
  metadata?: Record<string, any>;
  claimCount?: number;
  [key: string]: any;
}

export interface EvidenceEdgeData {
  id: string;
  source: string;
  target: string;
  label: "works_at" | "created" | "attended" | "contributed_to" | "published" | "profile_of" | "associated_with" | string;
  status: "verified" | "ambiguous" | "conflicting" | "normal" | string;
  sourceRef?: string;
  confidence?: string;
}

export interface EvidenceDetail {
  id: string;
  claim: string;
  source: string;
  sourceType: string;
  evidenceSnippet: string;
  reliability: "HIGH" | "MEDIUM" | "LOW" | string;
  timestamp: string;
  confidenceLevel: string;
  provenance: string;
  supportingSignals: string[];
  conflictingEvidence: string[];
  url?: string;
}

export interface SourceItem {
  id: string;
  source: string;
  platform: string;
  type: "Professional" | "Technical" | "Company" | "Event" | "Publication" | "Media" | string;
  reliability: "HIGH" | "MEDIUM" | "LOW" | string;
  lastChecked: string;
  evidenceCount: number;
  status: "VERIFIED" | "CONFLICTING" | "UNRESOLVED" | "ARCHIVED" | string;
  endpointOrDomain: string;
  verifiedSignatures: number;
}

export interface InvestigationCoverage {
  category: "Professional" | "Technical" | "Social" | "Events" | "Projects" | "Publications" | string;
  coveragePercent: number;
  evidenceCount: number;
  note: string;
}

export interface AuditTrailStep {
  id: string;
  step_index: number;
  stage: string;
  agent: string;
  agent_role: string;
  action: string;
  source: string;
  claim: string;
  result: string;
  timestamp: string;
  status: string;
  provenance?: string;
}

export interface AgentRecord {
  id: string;
  name: string;
  role: string;
  permissions: string[];
  constraints: string;
  avatar_icon?: string;
  status: string;
  actions_performed: string[];
  findings_count: number;
}

export interface RiskUncertainty {
  evidence_status: string;
  supporting_evidence_count: number;
  contradicting_evidence_count: number;
  unresolved_signals_count: number;
  major_concerns: string[];
  human_review_required: boolean;
  risk_level: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" | string;
}

export interface HumanReviewState {
  status: string;
  analyst_name?: string | null;
  analyst_decision: string;
  accepted_evidence_ids: string[];
  rejected_evidence_ids: string[];
  analyst_notes: string;
  signed_off: boolean;
  sign_off_timestamp?: string | null;
}

export interface Investigation {
  id: string;
  subjectName: string;
  alias: string;
  organization: string;
  knownPlatform: string;
  avatarUrl: string;
  createdAt: string;
  status: InvestigationStatus;
  coverage: InvestigationCoverage[];
  pipeline: PipelineStep[];
  dnaSignals: SignalMetric[];
  candidates: CandidateIdentity[];
  twinGuard: TwinGuardComparison;
  counterfactual: CounterfactualAnalysis;
  timeline: TimelineEventItem[];
  aiAnalyst: AIAnalystReport;
  sourcesCount: number;
  candidatesCount: number;
  evidenceLinksCount: number;
  ambiguitiesCount: number;
  nodes: EvidenceNodeData[];
  edges: EvidenceEdgeData[];
  // Advanced Investigation Engine Features
  audit_trail?: AuditTrailStep[];
  agents?: AgentRecord[];
  risk_uncertainty?: RiskUncertainty;
  human_review?: HumanReviewState;
}

export interface AdversarialTestCase {
  id: string;
  title: string;
  category:
    | "Normal Identity"
    | "Common Name"
    | "Similar Username"
    | "Different Aliases"
    | "Similar Profiles"
    | "Look-Alike"
    | "Conflicting Information"
    | "Conflicting Organization"
    | "Conflicting Timeline"
    | "Missing Information"
    | "Outdated Information"
    | "Insufficient Evidence"
    | string;
  inputContext: {
    name: string;
    alias?: string;
    org?: string;
    photoType: string;
  };
  candidatesFound: number;
  detectedRisk: string;
  finalStatus: ConfidenceStatus;
  systemReasoning: string;
  twinGuardTriggered: boolean;
  counterfactualSummary: string;
  evidenceStrength?: number;
  majorConcerns?: string[];
  supportingCount?: number;
  contradictingCount?: number;
  selfChallengeResult?: string;
}

export interface DiscoveredProfile {
  profile_id: string;
  platform: string;
  display_name: string;
  username: string;
  url: string;
  source_type: "PROFESSIONAL" | "COMMUNITY" | "TECHNICAL" | "EDUCATIONAL" | string;
  category: "PROFESSIONAL" | "COMMUNITY" | "TECHNICAL" | "EDUCATIONAL" | string;
  description: string;
  matched_signals: string[];
  evidence: string[];
  reliability: "HIGH" | "MEDIUM" | "LOW" | string;
  status: "DISCOVERED" | "NOT_DISCOVERED" | string;
  retrieved_at: string;
}

export interface PublicRecordItem {
  id: string;
  title: string;
  category: "COMMUNITY" | "TECHNICAL" | "EDUCATIONAL" | "PROJECTS" | "EVENTS" | string;
  source: string;
  url: string;
  evidence: string;
  retrieved_at: string;
  reliability: "HIGH" | "MEDIUM" | "LOW" | string;
}

export interface ProfileDiscoveryReport {
  id: string;
  subject_name: string;
  alias: string;
  organization: string;
  domain: string;
  additional_context: string;
  avatar_url: string;
  created_at: string;
  status: string;
  queries_generated: string[];
  sources_searched_count: number;
  profiles_discovered_count: number;
  verified_sources_count: number;
  profiles: DiscoveredProfile[];
  public_records: PublicRecordItem[];
  categories: {
    PROFESSIONAL?: DiscoveredProfile[];
    COMMUNITY?: DiscoveredProfile[];
    TECHNICAL?: DiscoveredProfile[];
    EDUCATIONAL?: DiscoveredProfile[];
    PROJECTS?: DiscoveredProfile[];
    EVENTS?: DiscoveredProfile[];
    [key: string]: DiscoveredProfile[] | undefined;
  };
  confidence_assessment?: string;
  why_this_result?: string[];
  discovery_summary: string;
  disclaimer: string;
}

export interface DiscoveryProgressState {
  discovery_id: string;
  status: "IN_PROGRESS" | "COMPLETED";
  current_stage: string;
  progress_percent: number;
  completed_stages: string[];
  queries_generated: number;
  sources_searched: number;
  profiles_discovered: number;
  community_records: number;
  project_records: number;
  event_records: number;
  verified_sources: number;
  result?: ProfileDiscoveryReport | null;
}
